File: backend/Margorak.Api/Dto/MapInteractionDto.cs
using System.Text.Json.Serialization;

namespace Margorak.Api.Dto
{

    [JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
    [JsonDerivedType(typeof(ShopInteractionDto), "shop")]
    [JsonDerivedType(typeof(TeleporterInteractionDto), "teleporter")]
    public abstract class MapInteractionDto
    {
        public int Id { get; set; }
        public string Description { get; set; } = string.Empty;
    }

    public class ShopInteractionDto :MapInteractionDto
    {
        public string ShopName { get; set; } = string.Empty;
    }

    public class TeleporterInteractionDto : MapInteractionDto
    {
        public int DestinationMapId {  get; set; }
        public int DestinationLocX { get; set; }
        public int DestinationLocY { get; set; }

    }
}

File: frontend/src/app/features/map/dto/map-interaction.dto.ts
interface BaseInteractionDto {
  id: number;
  type: string;
  description: string;
}

export interface ShopInteractionDto extends BaseInteractionDto {
  type: 'shop';
  shopName: string;
}

export interface EncounterInteractionDto extends BaseInteractionDto {
  type: 'encounter';
}
export interface TeleporterInteractionDto extends BaseInteractionDto {
  type: 'teleporter';
  destinationMapId: number;
  destinationLocX: number;
  destinationLocY: number;
}

export type MapInteractionDto =
  ShopInteractionDto | TeleporterInteractionDto | EncounterInteractionDto;

export interface MapInteractionHandler {
  type: MapInteractionDto['type'];
  handle(interaction: MapInteractionDto): void;
}
